src/snowflake/cli/api/metrics.py

```diff
@@ -86,7 +207,109 @@ def increment_counter(self, name: str, value: int = 1) -> None:
         else:
             self._counters[name] += value
 
+    @property
+    def _current_step(self) -> Optional[_CLIMetricsStep]:
+        return self._executing_steps[-1] if len(self._executing_steps) > 0 else None
+
+    @contextmanager
+    def track_step(self, name: str):
+        """
+        Recommended general use API for tracking a step throughout the execution of a command
+        Assumes that parent steps contain the entirety of their child steps
+
+        start_step and end_step are exposed in case the caller needs more flexibility
+
+        :raises CliMetricsInvalidUsageError: if the step name is empty
+        """
+        step_id = self.start_step(name)
+        try:
+            yield
+        except BaseException as err:
+            self.end_step(step_id=step_id, error=err)
+            raise
+        else:
+            self.end_step(step_id=step_id)
+
+    def start_step(self, name: str) -> int:
+        """
+        Command for manually tracking steps throughout the execution of a command
+        E.g. through different scopes/files for when a context manager is not enough
+        Requires a corresponding call to end_step with the (optional) exception caught
+
+        :returns: id of the started step which can be passed to end_step
+
+        :raises CliMetricsInvalidUsageError: if the step name is empty
+        """
+        parent_step = self._current_step
+        new_step = _CLIMetricsStep(name, parent_step)
+        self._executing_steps.append(new_step)
+        new_step.start()
+        return new_step.step_id
+
+    def end_step(
```

Comment:
have we considered modelling this as a self-contained Step structure, such that start/stop would be instance methods of a step? It would fit nicely with the context manager abstraction as well. I think it would help simplify the implementation here quite a bit too.

Reply:
Discussed offline: we yield the span here in case we want to perform operations with it in the future, however for now we're not supporting manual start/stop and only supporting through context managers